### PR TITLE
fix the problem #5763

### LINKED ISSUE #5763: DROP OWNED BY fails to drop the schemas on the workers
```SQL
CREATE USER u1 WITH superuser;
SELECT run_command_on_workers($$CREATE USER u1 WITH superuser;$$);
\c - u1 - 
create schema test_sc;

DROP OWNED BY u1;

create schema test_sc;
ERROR:  schema "test_sc" already exists
CONTEXT:  while executing command on localhost:9700
```

### PR DESCRIPTION
**Pull request by TsinghuaLucky912**
DESCRIPTION: Fixes a bug that caused DROP OWNED BY to not drop schemas on workers

Dear developers, I have submitted a version of the code to solve the problem that the DROP OWNED BY command is not delivered to the worker node, which causes the failure to create objects such as schemas again.

1. link: https://github.com/citusdata/citus/issues/5763
2. fix the problem #5763 DROP OWNED BY fails to drop the schemas on the workers

Fixes #5763

### CODE REVIEW

**Comment by TsinghuaLucky912** [comment]
If this is allowed, I will ban the following content in the regression test. I want to know what everyone thinks of this problem?

```
SELECT run_command_on_workers($$DROP OWNED BY read_access$$);
                        run_command_on_workers                        
 ---------------------------------------------------------------------
- (localhost,57637,t,"DROP OWNED")
- (localhost,57638,t,"DROP OWNED")
+ (localhost,57637,f,"ERROR:  operation is not allowed on this node")
+ (localhost,57638,f,"ERROR:  operation is not allowed on this node")
 (2 rows)
```

image

File: src/backend/distributed/deparser/deparse_owned_stmts.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	foreach(cell, roleList)
+	{
+		Node *roleNode = (Node *) lfirst(cell);
+		Assert(IsA(roleNode, RoleSpec) || IsA(roleNode, AccessPriv));
+		char const *rolename = NULL;
+		if (IsA(roleNode, RoleSpec))
+		{
+			rolename = RoleSpecString((RoleSpec *) roleNode, true);
+		}
+		if (IsA(roleNode, AccessPriv))
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_owned_stmts.c:77]: Can you really have this? Postgres code seems to deal with only RoleSpec: https://github.com/postgres/postgres/blob/3d14e171e9e2236139e8976f3309a588bcc8683b/src/backend/commands/user.c#L1467

File: src/backend/distributed/deparser/deparse_owned_stmts.c [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		{
+			rolename = quote_identifier(((AccessPriv *) roleNode)->priv_name);
+		}
+		appendStringInfoString(buf, rolename);
+		if (cell != list_tail(roleList))
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_owned_stmts.c:82]: We need to add a test where we have multiple roles to trigger this check

File: src/backend/distributed/deparser/deparse_owned_stmts.c [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+	AppendRoleList(buf, stmt->roles);
+
+	if (stmt->behavior == DROP_RESTRICT)
+	{
+		appendStringInfo(buf, " RESTRICT");
+	}
+	else if (stmt->behavior == DROP_CASCADE)
+	{
+		appendStringInfo(buf, " CASCADE");
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_owned_stmts.c:60]: We should have a test that covers CASCADE and RESTRICT options

File: src/backend/distributed/commands/owned.c [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+ * for the workers.
+ */
+List *
+PreprocessDropOwnedStmt(Node *node, const char *queryString,
+					   ProcessUtilityContext processUtilityContext)
+{
+	DropOwnedStmt *stmt = castNode(DropOwnedStmt, node);
+	List *allDropRoles = stmt->roles;
+
+	List *distributedDropRoles = FilterDistributedRoles(allDropRoles);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/owned.c:62]: Could use a test where we have a non-distributed role:
```SQL
set citus.enable_create_role_propagation to off;

-- this is now local-only role
CREATE ROLE ....
```

File: src/backend/distributed/commands/owned.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	if (list_length(distributedDropRoles) <= 0)
+	{
+		return NIL;
+	}
+
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
```
Comment (onderkalaci) [inline src/backend/distributed/commands/owned.c:71]: Do we need this as well? 

```C
	/* check creation against multi-statement transaction policy */
	if (!ShouldPropagateCreateInCoordinatedTransction())
	{
		return false;
	}

```

**Review by onderkalaci — COMMENTED** [review]
Thanks for the contribution!

```
 DROP SCHEMA single_node_ent CASCADE;
 DROP OWNED BY full_access_single_node;
 DROP OWNED BY read_access_single_node;
 SELECT run_command_on_workers($$DROP OWNED BY full_access_single_node$$);
                        run_command_on_workers                        
 ---------------------------------------------------------------------
- (localhost,57637,t,"DROP OWNED")
+ (localhost,57637,f,"ERROR:  operation is not allowed on this node")
 (1 row)
```

I think you should also adjust the existing regression tests NOT to have ` SELECT run_command_on_workers($$DROP OWNED BY  XXX$$);` anymore

**Comment by TsinghuaLucky912** [comment]
@onderkalaci Thank you very much for your review, which is very helpful to me

**Comment by TsinghuaLucky912** [comment]
@onderkalaci 

Hi, I would like to ask this ci/circleci: check-style — Your tests failed on CircleCI regression testing question:
```
FAIL: src/include/distributed/deparser.h (File size changed from 10821 to 10822)
FAIL: src/include/distributed/commands.h (File size changed from 37021 to 37033)
FAIL: src/backend/distributed/commands/owned.c (File size changed from 2503 to 2501)
```
This is the first time I submit code for citus, not sure what this means and how to solve it?

I took a look at the actual size of the file, but couldn't see why：

image

**Comment by SaitTalhaNisanci** [comment]
It means your code doesn't comply with the style standards of the project, you can follow this to fix it:  https://github.com/citusdata/citus/blob/main/CONTRIBUTING.md#following-our-coding-conventions

**Comment by TsinghuaLucky912** [comment]
@SaitTalhaNisanci @onderkalaci Thank you so much for your documentation help, I performed a
Make reindent can be, CI has all passed Thank you again for your help!

**Comment by TsinghuaLucky912** [comment]
@SaitTalhaNisanci @marcocitus @onderkalaci Hello, I just updated my branch (update main) and caused some regression tests to fail, but I found no exceptions on the local environment (pg15), can you help me start CI again? thank you very much!


image

**Review by marcocitus — APPROVED** [review]

**Pull request merged**